### PR TITLE
feat: miner resolves hub addr via locator

### PR DESCRIPTION
**Pull request by oopcode**
* If hub's address is just an ethereum addr, miner looks up hub's addr via locator.
* Addresses returned by locator are checked for reachability, the first reachable address is used.

### CODE REVIEW

**Comment by oopcode** [comment]
```
$ python -c "import random; print(random.choice(['3Hren', '3Hren', '3Hren']))"
3Hren
```

File: insonmnia/miner/builder.go [hunk 1]
```diff
@@ -223,15 +229,71 @@ func (b *MinerBuilder) getPublicIPs() (pubIPs []string, err error) {
...
+
+		log.G(b.ctx).Info("resolved hub endpoints", zap.Any("endpoints", resolved.IpAddr))
+
+		for _, addr := range resolved.IpAddr {
+			addr = strings.Replace(addr, "10001", "10002", 1)
+
+			log.G(b.ctx).Debug("trying hub endpoint", zap.Any("endpoint", addr))
+
+			dialer := net.Dialer{DualStack: true, Timeout: time.Second}
+			testCC, err := dialer.DialContext(b.ctx, "tcp", addr)
```
Comment (sshaman1101) [inline insonmnia/miner/builder.go]: defer `testCC.Close()`

File: insonmnia/miner/builder.go [hunk 2]
```diff
@@ -223,15 +229,71 @@ func (b *MinerBuilder) getPublicIPs() (pubIPs []string, err error) {
...
+			if err != nil {
+				log.G(b.ctx).Debug(
+					"hub endpoint is unreachable", zap.Any("endpoint", addr),
+					zap.Any("error", err))
+			} else {
+				testCC.Close()
+				hubSockaddr, hubEthAddr, err := util.ParseEndpoint(hubEndpoint + "@" + addr)
+				if err == nil {
```
Comment (sshaman1101) [inline insonmnia/miner/builder.go]: what error will be returned if err is not nil?

File: insonmnia/miner/builder.go (hunk 1 above)
Comment (oopcode) [reply insonmnia/miner/builder.go]: It is closed if error is not nil (see below); do we expect an open connection if there was an error?

File: insonmnia/miner/builder.go (hunk 2 above)
Comment (oopcode) [reply insonmnia/miner/builder.go]: If `err` is not `nil` (which is sort of impossible in this case, but it's just a safety check) we move on to the next resolved address, and if none of the resolved addresses worked, we return an error.

Comment (oopcode) [reply insonmnia/miner/builder.go]: Oh, I get it, I'll make it return the most relevant error.

File: insonmnia/miner/builder.go [hunk 3]
```diff
@@ -223,15 +229,71 @@ func (b *MinerBuilder) getPublicIPs() (pubIPs []string, err error) {
...
+func NewMinerBuilder(cfg Config, key *ecdsa.PrivateKey) (*MinerBuilder, error) {
+	b := &MinerBuilder{key: key}
 	b.Config(cfg)
-	return b
+
+	if b.ctx == nil {
+		b.ctx = context.Background()
+	}
+
+	_, TLSConf, err := util.NewHitlessCertRotator(b.ctx, b.key)
```
Comment (noxiouz) [inline insonmnia/miner/builder.go]: do not drop rotator. Its method Stop should be called

Comment (oopcode) [reply insonmnia/miner/builder.go]: It's actually dropped in a couple of other places around the project. Should I fix those spots?

File: insonmnia/miner/builder.go [hunk 4]
```diff
@@ -223,15 +229,77 @@ func (b *MinerBuilder) getPublicIPs() (pubIPs []string, err error) {
...
+		resolved, err := b.locatorClient.Resolve(b.ctx, &pb.ResolveRequest{EthAddr: hubEndpoint})
+		if err != nil {
+			return "", common.Address{}, fmt.Errorf(
+				"failed to resolve hub addr from %s: %s", hubEndpoint, err)
+		}
+
+		log.G(b.ctx).Info("resolved hub endpoints", zap.Any("endpoints", resolved.IpAddr))
+
+		for _, addr := range resolved.IpAddr {
+			addr = strings.Replace(addr, "10001", "10002", 1)
```
Comment (3Hren) [inline insonmnia/miner/builder.go]: What the hell is it?

**Comment by oopcode** [comment]
`grpc.Balancer` is scheduled for the next sprint (paired with miner builder rewrite).

**Pull request merged**